rust/src/measurements/make_private_lazyframe/group_by/pseudocode/make_private_group_by.py

```diff
@@ -0,0 +1,101 @@
+# type: ignore
+def make_private_group_by(
+    input_domain,
+    input_metric,
+    output_measure,
+    plan,
+    global_scale,
+    threshold,
+):
+    input_, keys, aggs, expr_threshold = match_group_by(plan)
+
+    t_prior = input_.make_stable(input_domain, input_metric)
+    middle_domain, middle_metric = t_prior.output_space()
+
+    grouping_columns = match_grouping_columns(keys)
+
+    margin = middle_domain \
+        .margins \
+        .get(grouping_columns, Margin.default())
+
+    # prepare a join measurement over all expressions
+    expr_domain = ExprDomain(
+        middle_domain,
+        ExprContext.Aggregate(grouping_columns),
+    )
+
+    m_exprs = make_basic_composition([
+        make_private_expr(
+             expr_domain,
+             PartitionDistance(middle_metric),
+             output_measure,
+             expr,
+             global_scale,
+        ) for expr in aggs
+    ])
+
+    # reconcile information about the threshold |\label{reconcile-threshold}|
+    dp_exprs = m_exprs.invoke((input_, all()))
+
+    if threshold is not None and expr_threshold is not None:
+        raise "two thresholds set"
```

Comment:
I dropped this validation, so that people can still filter outputs through the context API, even when not necessary for privacy. This can give some nice computational benefits.